request_hooks.go

```diff
@@ -0,0 +1,60 @@
+// Copyright (c) HashiCorp, Inc.
+// SPDX-License-Identifier: MPL-2.0
+
+package tfe
+
+import (
+	"context"
+	"net/http"
+)
+
+// ContextWithResponseHeaderHook returns a context that will, if passed to
+// [ClientRequest.Do] or to any of the wrapper methods that call it, arrange
+// for the given callback to be called with the headers from the raw HTTP
+// response.
+//
+// This is intended for allowing callers to respond to out-of-band metadata
+// such as cache-control-related headers, rate limiting headers, etc. Hooks
+// must not modify the given [http.Header] or otherwise attempt to change how
+// the response is handled by [ClientRequest.Do]; instead, they should just
+// save the provided header object and then interact with it only after
+// [ClientRequest.Do] has returned.
+//
+// If the given context already has a response header hook then the returned
+// context will call both the existing hook and the newly-provided one, with
+// the newer being called first.
+func ContextWithResponseHeaderHook(parentCtx context.Context, cb func(status int, header http.Header)) context.Context {
+	// If the given context already has a notification callback then we'll
+	// arrange to notify both the previous and the new one. This is not
+	// a super efficient way to achieve that but we expect it to be rare
+	// for there to be more than one or two hooks associated with a particular
+	// request, so it's not warranted to optimize this further.
+	existingI := parentCtx.Value(contextResponseHeaderHookKey)
+	finalCb := cb
+	if existingI != nil {
+		existing := existingI.(func(int, http.Header))
```

Comment:
This was the failing errcheck. I'm not sure why this isn't showing up in CI, so I'll check into that.

```
$ make lint
golangci-lint run .
request_hooks.go:35:15: Error return value is not checked (errcheck)
		existing := existingI.(func(int, http.Header))
```

Reply:
:thinking: interesting...

So is the linter rule here that it's just totally forbidden to use the one-result form of type assertion?

It looks like `golangci-lint` is wrapping [`errcheck`](https://github.com/kisielk/errcheck/blob/8f86be0cd56a95f4b0ddcbab3eec22f04e58b202/errcheck/errcheck.go#L620-L635) here and indeed it does seem to have [a rather-onerous rule that single-result type assertions are just never allowed](https://github.com/kisielk/errcheck/blob/8f86be0cd56a95f4b0ddcbab3eec22f04e58b202/errcheck/errcheck.go#L620-L635).

The only way this assertion could fail is if another part of this file were buggy, so I'm going to try to resolve this by just _explicitly_ panicking if the assertion fails, thereby achieving exactly the same effect as what I originally wrote but hopefully in a way that will satisfy the linter.